onboarding/mlsys/2-autodiff.md

```diff
@@ -0,0 +1,399 @@
+function $h$ composed of functions $f$ and $g$:
+
+```math
+h(x) = f(g(x))
+```
+
+then the derivative of $h$ is:
+
+```math
+h'(x) = f'(g(x)) \cdot g'(x)
+```
+
+The chain rule applies to derivatives of nested functions.  If we think of
+successive layers of a model as nested functions, then we can imagine the
+following:
+
+- $x$ is the input to the model
+- $g$ is the composition of all previous layers of the model
+- $g(x)$ is the output of the previous layer
+- $f$ is the current layer of the model
+- $h$ is the composition of all previous layers and also the current one
+
+Now let's replace the term $f'(g(x))$ with an arbitrary variable name $Z$:
+
+```math
+h'(x) = Z \cdot g'(x)
+```
+
+We can say that $Z$ must be a value that, multiplied by the gradient of previous
+layers ($g'(x)$), results in the gradient including the current layer ($h'(x)$).
+
+If $Z$ were particularly easy to calculate, it would be a way to calculate the
+total gradient incrementally, by calculating the equivalent $Z$ term for each
+layer and multiplying them together.  But is $Z$ easy to calculate for a layer?
+
+The answer turns out to be "yes".  $Z$ is the result of feeding the output of
+the previous layer $g(x)$, which is easy to compute, into $f'$, the derivative
+of the operator $f$, which is often trivial to obtain (for example, if $f(n) =
+n^2$ then $f'(n) = 2n$).
+
+And so if we break a model into individual operations and track the output of
+each layer, then we can express what contribution each layer makes to the total
+gradient.
+
+### A Complete Example, By Hand
+
+Let's remember our loss function from the previous exercise:
+
+```math
+\mathrm{loss_point} = (m x - \mathrm{ytrue})^2
+```
+
+We can think of the output as a sequence of three operations applied to a
+starting value $x$:
+
+- multiply by $m$
+- subtract $ytrue$
+- square the result
+
+### Q1
+
+Fill in the **output** and **$f'(g(x))$** columns of the following table.
+Remember when calculating the output column that the input to each layer is
+the output of the previous layer.  Also remember that `g(x)` just refers to the
+output of the previous layer.
+
+following table representing the forward pass (that is, where you
+calculate outputs of each sequential layer).  Let `x = 3`, `ytrue
+= 5`, and `m = 7`.
+
+| layer                 |  `f(n)`        | output | `f'(n)` | `f'(g(x))` |
+| --------------------- | -------------- | ------ | ------- | ---------- |
+| start with $x = 3$    |   n/a          |   1    |   n/a   |    n/a     |
+| multiply by $m = 7$   |  $m \cdot n$   |        |   $m$   |            |
+| subtract $ytrue = 5$  |  $n-5$         |        |   $1$   |            |
+| square the result     |  $n^2$         |        |   $2n$  |            |
+
+And lastly, what is the product of all the numbers in the final column?
+
+### A Complete Example, In PyTorch
+
+Now let's have PyTorch do the same thing for us.
+
+Start with some tensors `x` and `y`.
+
+```python
```

Comment:
I understand you're going for pedagogy here, but it would be very unusual to set `requires_grad = True` on `x` or `y`, or on any tensor which is not going to be updated by the optimizer, since this would just lead to unnecessary caching and computation.  Maybe just a comment to that effect is enough?